File: app/src/main/java/app/fit/fitndflow/domain/usecase/AddExerciseUseCase.java
package app.fit.fitndflow.domain.usecase;

import static app.fit.fitndflow.domain.Utils.convertToStringInLanguages;

import android.content.Context;

import java.util.List;

import app.fit.fitndflow.data.common.SharedPrefs;
import app.fit.fitndflow.domain.common.usecase.FitUseCase;
import app.fit.fitndflow.domain.model.ExerciseModel;
import app.fit.fitndflow.domain.model.ExerciseModelInLanguages;
import app.fit.fitndflow.domain.repository.FitnFlowRepository;
import io.reactivex.Single;

public class AddExerciseUseCase extends FitUseCase<ExerciseModelInLanguages, List<ExerciseModel>> {

    private FitnFlowRepository fitnFlowRepository;
    private Context context;

    public AddExerciseUseCase(Context context, String language, String nameExercise, int categoryId, FitnFlowRepository fitnFlowRepository){
        super(new ExerciseModelInLanguages(-1, convertToStringInLanguages(language, nameExercise), categoryId));
        this.context = context;
        this.fitnFlowRepository = fitnFlowRepository;
    }
    @Override
    public Single<List<ExerciseModel>> buildUseCaseObservable() {
        return Single.create(emitter -> {
            try{
            String apiKey = SharedPrefs.getApikeyFromSharedPRefs(context);
            List<ExerciseModel> response = fitnFlowRepository.addNewExercise(inputParams.getName(), inputParams.getCategoryId(), apiKey);
            emitter.onSuccess(response);
        } catch (Exception exception) {
            emitter.onError(exception);
        }
        });
    }
}

File: app/src/main/java/app/fit/fitndflow/domain/usecase/AddExerciseUseCase.kt
package app.fit.fitndflow.domain.usecase

import android.content.Context
import app.fit.fitndflow.data.common.SharedPrefs
import app.fit.fitndflow.domain.common.usecase.UseCase
import app.fit.fitndflow.domain.model.ExerciseModel
import app.fit.fitndflow.domain.repository.FitnFlowRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class AddExerciseUseCase(val fitnFlowRepository: FitnFlowRepository, val context: Context) : UseCase<AddExerciseUseCaseParams, List<ExerciseModel>>() {
    override fun run(params: AddExerciseUseCaseParams): Flow<List<ExerciseModel>>  = flow {
        val apiKey = SharedPrefs.getApikeyFromSharedPRefs(context)
        val newExercise = fitnFlowRepository.addNewExercise(params.name, params.language, params.categoryId, apiKey)
        emit(newExercise)
    }
}
data class AddExerciseUseCaseParams(var name : String, val language: String, var categoryId : Int)
